File: src/main/scala/com/github/neo4j/repository/Neo4jConnector.scala
package com.github.neo4j.repository

import org.neo4j.driver.Logging
import org.neo4j.ogm.config.Configuration
import org.neo4j.ogm.drivers.bolt.driver.BoltDriver
import org.neo4j.ogm.drivers.bolt.driver.BoltDriver.CONFIG_PARAMETER_BOLT_LOGGING
import org.neo4j.ogm.session.{Neo4jSession, Session, SessionFactory}
import org.neo4j.ogm.transaction.Transaction
import org.slf4j.{Logger, LoggerFactory}

import java.text.SimpleDateFormat
import java.util.Date

abstract class Neo4jConnector(
    url: String,
    username: String,
    password: String,
    database: String,
    packageName: String)
    extends AutoCloseable:

  private var _logger: Logger                = _
  private var driver: BoltDriver             = _
  private var sessionFactory: SessionFactory = _

  protected def build(): Neo4jConnector =
    this.driver = buildDriver(url, username, password, database)
    this.sessionFactory = buildSessionFactory(this.driver, packageName)
    this._logger = _buildLogger
    this

  private def _buildLogger: Logger = LoggerFactory.getLogger(getClass)

  private def buildDriver(url: String, username: String, password: String, database: String)
      : BoltDriver =
    val driver = new BoltDriver
    val configuration = new Configuration.Builder()
      .uri(url)
      .verifyConnection(true)
      .withCustomProperty(CONFIG_PARAMETER_BOLT_LOGGING, Logging.slf4j())
      .credentials(username, password)
      .database(database)
      .build
    driver.configure(configuration)
    driver

  private def buildSessionFactory(driver: BoltDriver, packageNames: String): SessionFactory =
    new SessionFactory(driver, packageNames)

  private def session: Session = this.sessionFactory.openSession.asInstanceOf[Neo4jSession]

  protected def withTransaction[R](function: Function[Session, R]): R =
    val session     = this.session
    val transaction = session.beginTransaction(Transaction.Type.READ_WRITE)
    val results     = function(session)
    transaction.commit()
    results

  protected def withSession[R](function: Function[Session, R]): R = function(this.session)

  override def close(): Unit = driver.close()
end Neo4jConnector

File: crud/src/main/java/io/github/abhishekghoshh/crud/neo4j/Neo4jConnector.java
package io.github.abhishekghoshh.crud.neo4j;

import ac.simons.neo4j.migrations.core.Migrations;
import ac.simons.neo4j.migrations.core.MigrationsConfig;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.neo4j.driver.AuthTokens;
import org.neo4j.driver.GraphDatabase;
import org.neo4j.driver.Logging;
import org.neo4j.ogm.config.Configuration;
import org.neo4j.ogm.drivers.bolt.driver.BoltDriver;
import org.neo4j.ogm.session.Session;
import org.neo4j.ogm.session.SessionFactory;
import org.neo4j.ogm.transaction.Transaction;

import java.util.function.Function;

public abstract class Neo4jConnector implements AutoCloseable {

    private static final Logger logger = LogManager.getLogger(Neo4jConnector.class);

    private final String url;
    private final String username;
    private final String password;
    private final String database;
    private final String packageName;
    private boolean applyMigration = false;


    private BoltDriver driver;
    private SessionFactory sessionFactory;

    public Neo4jConnector(String url, String username, String password, String database, String packageName) {
        this.url = url;
        this.username = username;
        this.password = password;
        this.database = database;
        this.packageName = packageName;
        build();
    }

    public Neo4jConnector applyMigrations() {
        this.applyMigration = true;
        return this;
    }

    private void startMigrations(String url, String username, String password, String database, String schemaDatabase) {
        var graphDatabaseDriver = GraphDatabase.driver(url, AuthTokens.basic(username, password));
        MigrationsConfig migrationConfig = MigrationsConfig.builder()
                .withLocationsToScan("classpath:db/migrations")
                .withDatabase(database)
                .withSchemaDatabase(schemaDatabase)
                .build();
        var migrations = new Migrations(migrationConfig, graphDatabaseDriver);

        migrations.apply(true).ifPresent(migrationVersion ->
                logger.info("migration version {}", migrationVersion.getValue())
        );
    }

    // Run after the bean has been created
    protected Neo4jConnector build() {
        this.driver = buildDriver(url, username, password, database);
        this.sessionFactory = buildSessionFactory(this.driver, packageName);
        logger.debug("migrations flag: {}", this.applyMigration);
        if (this.applyMigration) {
            logger.debug("Applying neo4j migrations");
            startMigrations(url, username, password, database, database);
        }
        return this;
    }

    private BoltDriver buildDriver(String url, String username, String password, String database) {
        BoltDriver driver = new BoltDriver();
        Configuration configuration = new Configuration.Builder()
                .uri(url)
                .verifyConnection(true)
                .withCustomProperty(BoltDriver.CONFIG_PARAMETER_BOLT_LOGGING, Logging.slf4j())
                .credentials(username, password)
                .database(database)
                .build();
        driver.configure(configuration);
        return driver;
    }

    private SessionFactory buildSessionFactory(BoltDriver driver, String packageName) {
        return new SessionFactory(driver, packageName);
    }

    private Session getSession() {
        return this.sessionFactory.openSession();
    }

    public <R> R withTransaction(Function<Session, R> function) {
        Session session = getSession();
        try (Transaction transaction = session.beginTransaction(Transaction.Type.READ_WRITE)) {
            R result = function.apply(session);
            transaction.commit();
            return result;
        }
    }

    public <R> R withSession(Function<Session, R> function) {
        return function.apply(getSession());
    }

    @Override
    public void close() {
        if (driver != null) {
            driver.close();
        }
    }
}
